bot/exts/moderation/clean.py

```diff
@@ -0,0 +1,598 @@
+        if not channels:
+            # Input was validated - if first_limit is a message, second_limit won't point at a different channel.
+            if isinstance(first_limit, Message):
+                channels = [first_limit.channel]
+            elif isinstance(second_limit, Message):
+                channels = [second_limit.channel]
+            else:
+                channels = [ctx.channel]
+
+        if isinstance(first_limit, Message):
+            first_limit = first_limit.created_at
+        if isinstance(second_limit, Message):
+            second_limit = second_limit.created_at
+        if first_limit and second_limit:
+            first_limit, second_limit = sorted([first_limit, second_limit])
+
+        # Needs to be called after standardizing the input.
+        predicate = self._build_predicate(bots_only, users, regex, first_limit, second_limit)
+
+        # Delete the invocation first
+        await self._delete_invocation(ctx)
+
+        if channels == "*" and use_cache:
+            message_mappings, message_ids = self._get_messages_from_cache(traverse=traverse, to_delete=predicate)
+        else:
+            deletion_channels = channels
+            if channels == "*":
+                deletion_channels = [channel for channel in ctx.guild.channels if isinstance(channel, TextChannel)]
+            message_mappings, message_ids = await self._get_messages_from_channels(
+                traverse=traverse,
+                channels=deletion_channels,
+                to_delete=predicate,
+                before=second_limit,
+                after=first_limit  # Remember first is the earlier datetime.
+            )
+
+        if not self.cleaning:
+            # Means that the cleaning was canceled
+            return
+
+        # Now let's delete the actual messages with purge.
+        self.mod_log.ignore(Event.message_delete, *message_ids)
+        deleted_messages = await self._delete_found(message_mappings)
+        self.cleaning = False
+
+        logged = await self._modlog_cleaned_messages(deleted_messages, channels, ctx)
+
+        if logged and is_mod_channel(ctx.channel):
+            with suppress(NotFound):  # Can happen if the invoker deleted their own messages.
+                await ctx.message.add_reaction(Emojis.check_mark)
+
+    # region: Commands
+
+    @group(invoke_without_command=True, name="clean", aliases=["clear", "purge"])
+    async def clean_group(
+            self,
+            ctx: Context,
+            users: Greedy[User] = None,
+            traverse: Optional[int] = None,
+            first_limit: Optional[CleanLimit] = None,
+            second_limit: Optional[CleanLimit] = None,
+            use_cache: Optional[bool] = None,
+            bots_only: Optional[bool] = False,
+            regex: Optional[Regex] = None,
+            *,
+            channels: CleanChannels = None  # "Optional" with discord.py silently ignores incorrect input.
+    ) -> None:
```

Comment:
Minor thing but I think the parameter list is indented 4 more spaces than it should be